### ISSUE TITLE
.save() is throwing duplicate key value violates unique constraint

### ISSUE THREAD

**Issue by DrakkenSaer** [typeorm/typeorm#4122]
**Issue type:**

[x] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[x] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

This was working for me earlier but now it is not. I am using a composite key of three different properties. Every time I try to update and existing row, I get the error `QueryFailedError: duplicate key value violates unique constraint "PK_8b1375852f33d30b328388e5a5c"`.

This is my entity:
```
'use strict';

import * as _ from 'lodash';
import { Entity, Column, PrimaryColumn, Generated } from "typeorm";


@Entity()
export class Assortment {

	@Column("uuid")
	@Generated("uuid")
	id!: string;

	@PrimaryColumn()
	accountId!: string;

	@PrimaryColumn()
	entityType!: string;

	@PrimaryColumn()
	parentId!: string;

	@Column({
		nullable: true
	})
	parentType!: string;

	@Column('varchar', {
		nullable: true
	})
	sortedIds!: string[];

	@Column('jsonb', {
		nullable: true
	})
	children!: Assortment[];
};
```

**Comment by aosifre** [comment]
I have exactly the same problem.

**Comment by PiiTan** [comment]
me three :(

**Comment by vamseekm** [comment]
Same kind of issue on postgres for me. I have a composite unique key (foreign key + a date column) throw error "QueryFailedError: duplicate key value violates unique constraint <unique index name>" when ever I use save method on this particular table.

**Comment by aosifre** [comment]
Still no patch concerning this duplication problem when backing up data?

**Comment by vogler** [comment]
I ran into the same issue in #3238. Using `find` + `insert/update` instead of just `save` is ugly.

**Comment by vogler** [comment]
As a workaround I added a PrimaryColumn id that is the concatenation of all the primary keys. Also ugly, but can be easily removed once this is fixed.

**Comment by ubershmekel** [comment]
`find + insert/update` was not working for me. Somehow the `find` was missing the instance. I don't know how. I tried out this upsert implementation from https://github.com/danielmhanover/typeorm-upsert

Which was great aside from:
* A small bug - it didn't use `"` quotes for the column names).
* I needed to provide the query builder from the connection, because I have multiple connections. This avoided the `Connection "default" was not found` error.

Here's my final workaround.

```

import { ObjectType, SelectQueryBuilder } from "typeorm";

/*
* entityType - TypeORM Entity
* obj - Object to upsert
* key_naming_transform (optional) - Transformation to apply to key names before upsert
* do_not_upsert - Keys to exclude from upsert. This is useful if a non-nullable field is required in case
* the row does not already exist but you do not want to overwrite this field if it already exists
*/
export async function upsert<T>(
  entityType: ObjectType<T>,
  obj: T,
  primary_key: string,
  queryBuilder: SelectQueryBuilder<any>,
  opts?: {
    key_naming_transform: (k: string) => string,
    do_not_upsert: string[],
  },
): Promise<T> {
  const keys: string[] = Object.keys(obj);
  for (const unkey of (opts ? opts.do_not_upsert : [])) {
    delete keys[unkey];
  }
  const setter_string =
    keys.map((k) => `"${opts ? opts.key_naming_transform(k) : k}" = :${k}`);

  const qb = queryBuilder
    .insert()
    .into(entityType)
    .values(obj)
    .onConflict(`("${primary_key}") DO UPDATE SET ${setter_string}`);

  keys.forEach((k) => {
    qb.setParameter(k, (obj as any)[k]);
  });

  return (await qb.returning('*').execute()).generatedMaps[0] as T;
}
```

**Comment by dekdekbaloo** [comment]
I can still confirm that this problem still exists. I'm using `typeorm` with `postgres`.

**Comment by ThCC** [comment]
> I can confirm that this problem still exists. I'm using `typeorm` with `postgres`.

Same here

**Comment by oscar-corredor** [comment]
same here

**Comment by NickKelly1** [comment]
Had a similar issue

Resolved it by converting all UUID's on save to lowercase.

MSSQL gave me UUID's in uppercase. Saving them to Postgres worked the first time since Postgres simply converts the UUID to lowercase.
On subsequent saves, TypeORM/Postgres is unable to match the lowercase and uppercase UUID's and tries to save as a new row, only to conflict with the existing row and throw the PK exception.

**Comment by bohendo** [comment]
Simple entity that this causing this problem with `typeorm@0.2.19` and postgres:
```
import { Column, Entity, PrimaryColumn } from 'typeorm';

@Entity('channel_records')
export class ChannelRecord {
  @PrimaryColumn()
  path!: string;

  @Column({ type: 'json' })
  value!: object;
}
```

This was working for a while but suddenly started throwing `QueryFailedError: duplicate key value violates unique constraint` after I restored a row from an external source. The `path` needs to be case sensitive so can't just blindly cast them to lower case.

**Comment by derSoerrn95** [comment]
Same problem with oracle. 

Using following entity and getting `Error: ORA-00001: Unique Constraint`, 'cause TypeORM wants to insert instead of update or in my case do nothing..

```typescript
export class DayData {

  @Column({name: 'DAY', type: 'date', primary: true})
  day: Date;

  @Column({name: 'ID1', type: 'number', primary: true})
  id1: number;

  @Column({name: 'ID2', type: 'number', primary: true})
  id2: number;

  @Column({name: 'QUANTITY'})
  quantity: number;
}
```

The problem is, that TypeORM is using a wrong date format during the select: 
```sql
SELECT 
    "DayData"."DAY" AS "DayData_DAY", 
    "DayData"."QUANTITY" AS "DayData_QUANTITY", 
    "DayData"."ID1" AS "DayData_ID1", 
    "DayData"."ID2" AS "DayData_ID2" 
FROM "DAY_DATA" "DayData" 
WHERE "DayData"."DAY" = :id_0_0 
    AND "DayData"."ID1" = :id_0_1 
    AND "DayData"."ID2" = :id_0_2 
;
-- PARAMETERS: ["2017-12-31T23:00:00.000Z",17,100]
```

And for the insert this one:
```sql
INSERT INTO "DAY_DATA"
    ("DAY", "QUANTITY", "ID1", "ID2") 
VALUES 
    (TO_DATE('2018-01-01', 'YYYY-MM-DD'), :i0_QUANTITY, :i0_ID1, :i0_ID2)
;
-- PARAMETERS: [18025,17,100]
```
It's a bit strange, that TypeORM itself casting the date param, but i think it's ok, because it uses the to_date function. But don't know if it's possible to get attacked by sql injection, there - need to test.

**Comment by n8sabes** [comment]
Having spent the last week building a platform on top of typeorm and then to discover there is no way to save an entity due to this bug is an extremely frustrating situation. I've tried working around this in various ways, without success.

Does anybody know of *ANY* solution to get Typeorm working with Postgres (using dual-foreign keys, not the single primary key uuid pattern)?

**Comment by duckies** [comment]
I noticed this today using just a simple primary generated key number and I cannot figure out how to update entities anymore. The update method is throwing the same error as the save method. This is a breaking issue that needs to be addressed asap.

**Comment by dulkith** [comment]
> Having spent the last week building a platform on top of typeorm and then to discover there is no way to save an entity due to this bug is an extremely frustrating situation. I've tried working around this in various ways, without success.
> 
> Does anybody know of _ANY_ solution to get Typeorm working with Postgres (using dual-foreign keys, not the single primary key uuid pattern)?

You find any solution for this?

**Comment by ubershmekel** [comment]
Using the upsert technique I mentioned above, I wanted to know whether a conflict occurred or not.

I had to replace this:

```
  return (await qb.returning('*').execute()).generatedMaps[0] as T;
```

With this:

```
  const executed = await qb.returning('*').execute();
  const wasCreated = !executed.raw[0].data;
  return executed.generatedMaps[0] as T;
```

But this will probably only correctly differentiate between if the entity model does not have `data` column. I also wonder if this would be considered a typeorm implementation detail or not.

@dulkith I think the only way to solve it is to manually craft a raw query or use a query builder. You might be able to adapt the snippet I adapted.

**Comment by lapwat** [comment]
This issue occurs when you are trying to save() an object with its relations.

Example with 2 tables: Photo and Category

```js
// first save
photo = { id: 1, category: { id: 1, name: "holidays" } }
this.save(photo) // works
this.save(photo) // does not work the second time
```

It does not work the second time because save() tries to recreate the relation between Photo(1) <=> Category(1).

You should do:

```js
// first save
photo = { id: 1, category: { id: 1, name: "holidays" } }
this.save(photo) // works
delete photo.category
this.save(photo) // works!
```

**Comment by saulotoledo** [comment]
I just got this problem, but I am not sure why it only happens if the entity has relationships for me.
In the meanwhile, try this: https://stackoverflow.com/a/21639138/2957291

It should solve the problem for now.

If I find more information before somebody else here I will paste updates.

**Comment by mehtamanan** [comment]
@saulotoledo Nope, it doesn't.

Still having troubles with this. Any updates on this? This seems like an important bug to fix.

**Comment by mehtamanan** [comment]
> This issue occurs when you are trying to save() an object with its relations.
> 
> Example with 2 tables: Photo and Category
> 
> ```js
> photo = { id: 1, category: { id: 1, name: "holidays" } }
> 
> // first save works
> this.save(photo) // works
> 
> // second save does not work
> this.save(photo)
> ```
> 
> It does not work the second time because save() tries to recreate the relation between Photo(1) <=> Category(1).
> 
> You should do:
> 
> ```js
> photo = { id: 1, category: { id: 1, name: "holidays" } }
> 
> // first save works
> this.save(photo)
> 
> delete photo.category
> 
> // second save works!
> this.save(photo)
> ```

@lapwat Are you saying that we cannot use .save to cascade updates down related entities, but we can use it to update just the base entity (and no related entities)?

**Comment by JosephBagaric** [comment]
> I just got this problem, but I am not sure why it only happens if the entity has relationships in my case.
> I believe you are using migrations with the QueryRunner or restoring a SQL dump.
> 
> In the meanwhile, try this after inserting the data in your database (in the end of your migration/seed, for example) **(PostgreSQL only)**:
> 
> ```sql
> SELECT setval(
>     pg_get_serial_sequence('"your_table_name"', 'your_primary_key_name'),
>     (SELECT MAX("id") FROM "your_table_name") + 1
> )
> ```
> 
> Source: https://stackoverflow.com/a/21639138/2957291
> 
> It should solve the problem for now.
> 
> If I find more information before somebody else here I will paste updates.
> 
> @ubershmekel @dulkith @n8sabes @bohendo @oscar-corredor @dekdekbaloo @vogler @aosifre @vamseekm @DrakkenSaer, can you confirm if this solves your problem?

This worked for me. Thanks.

**Comment by huantaoliu** [comment]
any updates on this? Still failling for me using postgres  with dual keys (2 columns) as primary key.

**Comment by Evilart86** [comment]
I just removed all objects from table, and then i saved my model without error. 
I tried to reproduce but it works good now..

**Comment by guillenotfound** [comment]
Any solution on this one? I have 1 PK and 1 UNIQUE and since I'm not passing the PK it will always attempt to insert, which is not desirable since it will raise `duplicate key value violates unique constraint`, are there any workarounds working?

**Comment by jhz7** [comment]
Hi guys. I was getting the same problem a few hours ago. I have discovered that if you don't tell entity id explicitly, typeorm is gonna try to create a new one. I have just initialized id (when it exists) and it solved problem.

This aproach was generating the issued error:

`@Entity('my_entity')
class MyEntity {

  constructor(domain: MyDomainObject) {
    if(domain) {
      this.title = domain.title;
      this.year = domain.year;
    }
  }

  @PrimaryGeneratedColumn()
  id!: number;

  @Column()
  title!: string;
  
  @Column()
  year!: number;
}`

But this one solved (at least for me):

`@Entity('my_entity')
class MyEntity {

  constructor(domain: MyDomainObject) {
    if(domain) {
      this.title = domain.title;
      this.year = domain.year;
    }

    if(domain.id) {
      this.id = domain.id;
    }
  }

  @PrimaryGeneratedColumn()
  id!: number;

  @Column()
  title!: string;
  
  @Column()
  year!: number;
}`

PD: 
- I'm using mysql 5 version.
- title and year togther, are one Unique constraint.

**Comment by goktugyil** [comment]
@jhz7 Can you format your code?

**Comment by Cerwym** [comment]
TypeORM : **0.2.22**

```
@Entity()
export class Device {
	@PrimaryColumn()
	public id: string;

	@Field()
	public name: string;


	@Field(() => [DeviceSetting], { nullable: true })
	@OneToMany(() => DeviceSetting, ds => ds.device, {
		eager: true,
		cascade: ['insert'],
	})
	public settings: DeviceSetting[];
}
```

```
@Entity()
@Index(['setting_id', 'device_id'], { unique: true })
export class DeviceSetting extends BaseEntity {
	@Field({
		name: 'id',
	})
	@PrimaryColumn()
	public setting_id: string;

	@PrimaryColumn()
	public device_id: string;

	@ManyToOne(() => Device, d => d.id)
	@JoinColumn({
		name: 'device_id',
	})
	public device!: Device;

	@ManyToOne(() => Setting, s => s.id, {
		eager: true,
	})
	@JoinColumn({
		name: 'setting_id',
		referencedColumnName: 'id',
	})
	public setting: Setting;
}
```

```
@Entity()
export class Setting {
	@PrimaryColumn({
		type: 'varchar',
		length: 14,
		unique: true,
		comment: 'The id code of a unique setting entry',
		select: false, // Marked as false because get device setting will generate ambiguous syntax with this selectable
	})
	public id: string;

	@Column()
	public description: string;
}
```

Indeed an issue, calling .Save on the **Device** entity will attempt to Insert DeviceSetting relation even if it was previously set in the model, rather than do nothing, it well attempt to insert the previous relation generating `duplicate key value violates unique constraint` as UQ `device_setting.device.id `& `sevice_setting.setting.id` already exists.

**Comment by Kritten** [comment]
I suddenly had the same problem and went down my git history.
At some point I switched the GraphQL type of my entity IDs from `Int` to `ID`. (My entity IDs are typed as `number` with autoincrement)
My Hotfix now is to manually parse the graphql ID (string) to a database ID (number).
Maybe I should have go with GUIDs from the beginning, but now it is too late

**Comment by bhanu-manthry** [comment]
> I can confirm that this problem still exists. I'm using `typeorm` with `postgres`.

Same here using `typeorm` with `mysql`.

**Comment by goktugyil** [comment]
> I just got this problem, but I am not sure why it only happens if the entity has relationships in my case.
> I believe you are using migrations with the QueryRunner or restoring a SQL dump.
> 
> In the meanwhile, try this after inserting the data in your database (in the end of your migration/seed, for example) **(PostgreSQL only)**:
> 
> ```sql
> SELECT setval(
>     pg_get_serial_sequence('"your_table_name"', 'your_primary_key_name'),
>     (SELECT MAX("id") FROM "your_table_name") + 1
> )
> ```
> 
> Source: https://stackoverflow.com/a/21639138/2957291
> 
> It should solve the problem for now.
> 
> If I find more information before somebody else here I will paste updates.
> 
> @ubershmekel @dulkith @n8sabes @bohendo @oscar-corredor @dekdekbaloo @vogler @aosifre @vamseekm @DrakkenSaer, can you confirm if this solves your problem?

worked for me

**Comment by JarLowrey** [comment]
I received this error when setting up seeding with TypeORM-Seeding. I fixed it by removing child creation from the factory methods of their parents (this might've caused and error as the parent wasn't created yet, not sure), but primarily I just forgot to call `await` on the factory methods in my seed file.

      let wcs = await factory(WristCounter)({user:user}).createMany(10);
      wcs.forEach(x => this.assignCounts(factory,x))

    async assignCounts(factory: Factory, x: WristCounter){
      x.datedCounts = await factory(WristCounterCount)({counter:x,user:x.userCreator}).createMany(10) as any;
    }

vs

      let wcs = await factory(WristCounter)({user:user}).createMany(10);
      wcs.forEach(x => x.datedCounts = factory(WristCounterCount)({counter:x,user:x.userCreator}).createMany(10) as any);

**Comment by imnotjames** [comment]
@JarLowrey seems unrelated to what's described in this thread.

Most of the examples here seem to be with multiple primary columns.  Is that the case?

If so, anyone want to confirm that #6417 fixes / doesn't fix the issue?

**Comment by DanielRamosAcosta** [comment]
Is there any workaround while a patch is being released? @imnotjames

**Comment by sshhawwnn** [comment]
This issue still persists. When I manually altered the row of records by exporting the whole table and import a new set of data into the table, this error occurs. It seems to me that TypeORM does not recognize the last inserted ID of the table? I have to run the script by @goktugyil in order to solve the issue.

**Comment by cohlar** [comment]
Same here, issue still persists when saving nested object with multiple keys in unique constraint - I use POSTGRES

**Comment by DanielRamosAcosta** [comment]
Also here, when saving an entity that have `OneToOne` reslation it throws `duplicate key value violates unique constraint`.

**Comment by DieserMerlin** [comment]
This happens for MongoDB nested objects too. Is there any workaround?

**Comment by DieserMerlin** [comment]
This is my workaround. Hope this helps someone.

Basically I just created a Repository Wrapper and use `extends SaveSafeRepository<User>` instead of `extends Repository<User>` to have a User repository that behaves as expected when you try to save an existing entity.

```typescript
import {DeepPartial, EntityRepository, Repository, SaveOptions, ObjectID} from 'typeorm';
import {IdEntity} from './id-entity';

@EntityRepository()
export class SaveSafeRepository<T> extends Repository<T> {
  save<T extends DeepPartial<T>>(
    entity: T & {_id?: ObjectID},
    options?: SaveOptions
  ): Promise<T & T> {
    //@ts-ignore
    if (!entity._id) return this.create(entity);
    //@ts-ignore
    else return this.update(entity._id, entity);
  }
}
```

Note that you possibly need to change the `_id` property to the name you use for IDs.

**Comment by cooperfrench95** [comment]
In case this helps at all. I had a situation where updating a one-to-one foreign key on a table would cause this error. What fixed it for me was simply creating a new entity rather than updating the existing one in order to get a new ID, then saving that new ID as the foreign key on the other table.

**Comment by odravison** [comment]
I'm facing the same issue here. Any update about fixing this wrong behavior ?

**Comment by mreppo** [comment]
check for existing record with findOne and save in case records does not exist, also this will prevent increasing autoincrement in case of you use maria|mysql, and InnoDB

**Comment by allencoded** [comment]
This is still not working for appropritely postgres.

**Comment by odravison** [comment]
I just figured out my scenario. In my case, I was trying to save 6 objects one by one with a supposed saved object inside each one. But when I debug I saw that my object inside that supposed to be saved, wasn't in some cases, this behavior only happened because I used concurrent save operation for each object. So I was getting this error always when attempt to save the second object.

**Comment by twigs67** [comment]
Is anyone from TypeOrm looking into this? This seems to be a huge issue.

**Comment by Dimitri3502** [comment]
I'm using a number as a primary key.
For me this error occurs when I try to save a list which includes duplicated items.
The method  save<T extends DeepPartial<Entity>>(entities: T[], options?: SaveOptions): Promise<(T & Entity)[]> throws the error "duplicate key value violates unique constraint".
The solution for me is to handle duplicates in my list before calling save()

**Comment by BaoPham92** [comment]
You could potentially just remove the ```unique``` options from the ```Decorators``` and still yet use the corresponding Column ```Decorators``` with your original logic for the time being.

**Comment by mattanimation** [comment]
wow, so I had this same issue and it turns out when searching [here](https://github.com/typeorm/typeorm/issues/6125) it dawned on me that it could be the same issue.
once I changed my **ConnectionOptions** to  **synchronize:false** it worked just fine.

**Comment by DedaDev** [comment]
Struggled with this for the last 5 hours, finally made it.

So first add a name to constraint in your entity,
```
@Entity()
@Unique('constraint_name', ['col_one', 'col_two'])
```

Then, you can use onConflict with ON CONSTRAINT

```
this.createQueryBuilder()
        .insert()
        .into(Score)
        .values(score)
        .onConflict(`ON CONSTRAINT constraint_name DO UPDATE SET whatever = 1`)
        .execute()
 );
```

**Comment by emagnier** [comment]
@imnotjames, thank you for your fix https://github.com/typeorm/typeorm/pull/6417. 

But in case you missed it (because this topic is now set at "closed"), this is still not working appropriately in Postgres.

**Comment by ryansonshine** [comment]
Can we re-open this? This is still an issue on Postgres as of [v0.2.31](https://github.com/typeorm/typeorm/releases/tag/0.2.31)

**Comment by jcrben** [comment]
@emagnier fyi this is working for me in postgres right now - I initially thought it wasn't but that was due to some bad data in my dev env

**Comment by RicardofVallejo** [comment]
I'm still facing the same issue in Postgres.
Current solution was to instead of using 

`  @PrimaryGeneratedColumn()
  id: number;`

I now use 

`  @PrimaryGeneratedColumn('uuid')
  id: number;`

It is not a fix but the odds of not having a matching Id helps

**Comment by sjohns21** [comment]
To help others in debugging: this happened to me because I inserted records using literal, numeric values (instead of DEFAULT or undefined) as arguments for the auto-incremented column. Doing so circumvents the column's [underlying sequence object'](https://www.postgresql.org/docs/8.2/functions-sequence.html)s increment call, hence making the sequence's value out of sync with the values in the column in the table.

**Comment by Form1ca** [comment]
still bug not fixed..

**Comment by jcrben** [comment]
@Form1ca can you provide a minimal reproduction?

**Comment by oscardianno** [comment]
> 
> 
> > I just got this problem, but I am not sure why it only happens if the entity has relationships in my case.
> > I believe you are using migrations with the QueryRunner or restoring a SQL dump.
> > In the meanwhile, try this after inserting the data in your database (in the end of your migration/seed, for example) **(PostgreSQL only)**:
> > ```sql
> > SELECT setval(
> >     pg_get_serial_sequence('"your_table_name"', 'your_primary_key_name'),
> >     (SELECT MAX("id") FROM "your_table_name") + 1
> > )
> > ```
> > 
> > 
> > Source: https://stackoverflow.com/a/21639138/2957291
> > It should solve the problem for now.
> > If I find more information before somebody else here I will paste updates.
> > @ubershmekel @dulkith @n8sabes @bohendo @oscar-corredor @dekdekbaloo @vogler @aosifre @vamseekm @DrakkenSaer, can you confirm if this solves your problem?
> 
> This worked for me. Thanks.

Excuse me, in this case, "your_table_name" is the name of the main table in both cases? The related table isn't meant to be included anywhere in the query?

**Comment by rafalebacalla** [comment]
I'm currently using Nest and SQLite, and the error is still there.
But I fixed it by manually changing the migration files. I found out that even though the `@Column()` already had `{ unique: false }`, its still unique in the migrations.

This is the non-unique field
![image](https://user-images.githubusercontent.com/58300147/135745848-5750876d-f61d-4485-a170-97b55e90f518.png)

But it generated unique in the migrations. I just removed this part of the code.
![image](https://user-images.githubusercontent.com/58300147/135745886-247cb273-ecab-4588-8508-a477d4ab4359.png)
